web-services-app/src/main/java/edu/unc/lib/boxc/web/services/processing/IiifV3ManifestService.java

```diff
@@ -0,0 +1,241 @@
+            throw new NotFoundException("No objects were found for inclusion in manifest for object " + pid.getId());
+        }
+        log.debug("Constructing manifest for {} containing {} items", pid.getId(), contentObjs.size());
+        ContentObjectRecord rootObj = contentObjs.get(0);
+        var manifest = new Manifest(getManifestPath(rootObj), new Label(getTitle(rootObj)));
+        manifest.setMetadata(buildMetadata(rootObj));
+        addAttribution(manifest, rootObj);
+
+        addItems(manifest, contentObjs);
+
+        return manifest;
+    }
+
+    private List<Metadata> buildMetadata(ContentObjectRecord rootObj) {
+        var metadataList = new ArrayList<Metadata>();
+        String abstractText = rootObj.getAbstractText();
+        if (abstractText != null) {
+            metadataList.add(new Metadata("description", abstractText));
+        }
+        addMetadataField(metadataList, "Creators", rootObj.getCreator());
+        addMetadataField(metadataList, "Subjects", rootObj.getSubject());
+        addMetadataField(metadataList, "Languages", rootObj.getLanguage());
+        metadataList.add(new Metadata("", "<a href=\"" +
+                URIUtil.join(baseAccessPath, "record", rootObj.getId()) + "\">View full record</a>"));
+        return metadataList;
+    }
+
+    private void addMetadataField(List<Metadata> metadataList, String fieldName, List<String> values) {
+        if (!CollectionUtils.isEmpty(values)) {
+            metadataList.add(new Metadata(fieldName, String.join(", ", values)));
+        }
+    }
+
+    private String getTitle(ContentObjectRecord contentObj) {
+        String title = contentObj.getTitle();
+        return (title != null) ? title : "";
+    }
+
+    private void addAttribution(Manifest manifest, ContentObjectRecord rootObj) {
+        String attribution = "University of North Carolina Libraries, Digital Collections Repository";
+        String collection = rootObj.getParentCollectionName();
+        if (collection != null) {
+            attribution += " - Part of " + collection;
+        }
+        manifest.setRequiredStatement(new RequiredStatement("Attribution", attribution));
+    }
+
+    /**
+     * Add canvas items for each record in the set being processed
+     * @param manifest
+     * @param contentObjs
+     */
+    private void addItems(Manifest manifest, List<ContentObjectRecord> contentObjs) {
+        var canvases = new ArrayList<Canvas>();
+        for (ContentObjectRecord contentObj : contentObjs) {
+            // Add canvases for any records with displayable content
+            if (hasViewableContent(contentObj)) {
+                canvases.add(createCanvas(contentObj));
+            }
+        }
+        manifest.setCanvases(canvases);
+    }
+
+    /**
+     * Constructs a standalone canvas document for the requested object
+     * @param pid
+     * @param agent
+     * @return
+     */
+    public Canvas buildCanvas(PID pid, AgentPrincipals agent) {
```

Comment:
not sure i get the difference between buildCanvas and createCanvas other than pid vs content object?

Reply:
One of them is public and is used for requests to build manifests by themselves rather than as a part of a manifest and wraps the other method, while the createCanvas one was used for building and populating the canvas so it could be used both by itself and as part of a manifest. I've renamed a number of methods in the class to try to make their purpose clearer.